ADBDEV-7258: Fix aborted database deletion on the entire cluster

RekGRpth opened the pull request:
Fix aborted database deletion on the entire cluster

Commit 034a9fc fixes aborted database deletion on a single instance, and this
patch fixes the same behavior on the entire GPDB cluster. The specified commit
first marked the deleted database as invalid using the heap_inplace_update
function, and only then did the actual database deletion using the irreversible
CatalogTupleDelete function, the call to which was moved further. This is not
enough to do for the entire cluster, because the GPDB first deletes the
database on the segments before deleting it on the coordinator. When resolving
conflicts with commit 45dce8b6, this logic was incorrectly left in place, which
led to an error on the entire cluster. This patch moves this logic immediately
before the call to the irreversible CatalogTupleDelete function.

Ticket: ADBDEV-7258

whitehawk commented on src/test/isolation2/sql/invalid_database.sql:
```diff
@@ -0,0 +1,45 @@
+DROP DATABASE IF EXISTS regression_invalid_interrupt;
+-- end_ignore
+CREATE TABLE store_session_id(a int, sess_id int);
+-- adding `2` as first column as the distribution column and add this tuple to segment 0
```
If the distribution is important for the test, maybe better to set distribution policy explicitly during table `store_session_id` creation?

RekGRpth replied:
simplified

whitehawk commented on src/test/isolation2/sql/invalid_database.sql:
```diff
@@ -0,0 +1,45 @@
+FROM gp_segment_configuration WHERE content = 0 AND role = 'p';
+
+-- and finally interrupt the DROP DATABASE on segment 0
+0U: SELECT pg_cancel_backend(pid) FROM pg_stat_activity JOIN store_session_id USING (sess_id)
+WHERE query = 'DROP DATABASE IF EXISTS regression_invalid_interrupt';
```
Why do we need to do JOIN here? Can we just do smth like:
```suggestion
0U: SELECT pg_cancel_backend(pid) FROM pg_stat_activity
WHERE query = 'DROP DATABASE IF EXISTS regression_invalid_interrupt';
```

RekGRpth replied:
simplified

whitehawk commented:
In the description please add more details about the root cause of the issue (why the mentioned commit is not enough for the GPDB cluster) and about the fix for GPDB.

RekGRpth commented:
> In the description please add more details about the root cause of the issue (why the mentioned commit is not enough for the GPDB cluster) and about the fix for GPDB.

added

hilltracer commented on src/backend/commands/dbcommands.c:
```diff
@@ -1132,6 +1132,9 @@ dropdb(const char *dbname, bool missing_ok)
 	RequestCheckpoint(CHECKPOINT_IMMEDIATE | CHECKPOINT_FORCE | CHECKPOINT_WAIT);
 
+
+	SIMPLE_FAULT_INJECTOR("dropdb_before_remove_tablespace");
```
Do we really need this fault injector? Why not use locks inside test instead?
Smth like this
<details>
<summary>diff</summary>

```diff
--- a/src/test/isolation2/sql/invalid_database.sql
+++ b/src/test/isolation2/sql/invalid_database.sql
@@ -8,21 +8,22 @@ DROP DATABASE IF EXISTS regression_invalid_interrupt;
 -- end_ignore
 CREATE DATABASE regression_invalid_interrupt;
 
--- prevent drop database via suspend in inject fault on segment 0
-SELECT gp_inject_fault('dropdb_before_remove_tablespace', 'suspend', dbid)
-FROM gp_segment_configuration WHERE content = 0 AND role = 'p';
+0U: BEGIN;
+0U: LOCK pg_tablespace;
 
 -- try to drop, this will wait due to the suspend on segment 0
 1&: DROP DATABASE regression_invalid_interrupt;
 
--- ensure we're suspending in inject fault on segment 0
-SELECT gp_wait_until_triggered_fault('dropdb_before_remove_tablespace', 1, dbid)
-FROM gp_segment_configuration WHERE content = 0 AND role = 'p';
+SELECT EXISTS (
+  SELECT 1 FROM pg_locks
+  WHERE NOT granted AND relation = 'pg_tablespace'::regclass
+  AND mode = 'AccessShareLock'
+);
 
--- and finally interrupt the DROP DATABASE on segment 0
-0U: SELECT pg_cancel_backend(pid) FROM pg_stat_activity
-WHERE query = 'DROP DATABASE IF EXISTS regression_invalid_interrupt';
-0Uq:
+0U: SELECT pg_cancel_backend(pid)
+     FROM pg_locks
+     WHERE NOT granted AND relation = 'pg_tablespace'::regclass
+     AND mode = 'AccessShareLock';
 
 1<:
 1q:
@@ -30,9 +31,8 @@ WHERE query = 'DROP DATABASE IF EXISTS regression_invalid_interrupt';
 -- verify that connection to the database aren't allowed
 ! psql -d regression_invalid_interrupt -c "SELECT 1";
 
--- to properly drop the database, we need to reset inject fault
-SELECT gp_inject_fault('dropdb_before_remove_tablespace', 'reset', dbid)
-FROM gp_segment_configuration WHERE content = 0 AND role = 'p';
+0U: ROLLBACK;
+0Uq:
 
 DROP DATABASE regression_invalid_interrupt;
 ```


</details>

RekGRpth replied:
applied

RekGRpth committed:
Fix aborted database deletion on the entire cluster

Commit 034a9fc fixes aborted database deletion on a single instance, and this
patch fixes the same behavior on the entire GPDB cluster. The specified commit
first marked the deleted database as invalid using the heap_inplace_update
function, and only then did the actual database deletion using the irreversible
CatalogTupleDelete function, the call to which was moved further. This is not
enough to do for the entire cluster, because the GPDB first deletes the
database on the segments before deleting it on the coordinator. When resolving
conflicts with commit 45dce8b, this logic was incorrectly left in place, which
led to an error on the entire cluster. This patch moves this logic immediately
before the call to the irreversible CatalogTupleDelete function.

RekGRpth commented:
bender build

RekGRpth committed:
fix

RekGRpth committed:
ignore
